test/gpu/native/studies/math/mathKernels.chpl

```diff
@@ -0,0 +1,52 @@
+use Math;
+import Time.stopwatch;
+use Random;
+
+config const size:uint(32) = 100_000_000;
+config const printTime = true;
+config const correctness = false;
+config param function = "tanhf";
+
+extern proc cu_nvcc_main(printTime: int, correctness: int): void;
+extern proc cu_clang_main(printTime: int, correctness: int): void;
+
+
+var D = {0..size};
+var CpuA : [D] real(32);
+
+proc main() {
+  fillRandom(CpuA);
+
+  on here.gpus[0]{
+    var A = CpuA;
+    const D = A.domain;
+
+    var s : stopwatch;
+    s.start();
+    kernel(function, D, A);
+    s.stop();
+    if printTime then
+      writeln("Chapel Time (", function, "): ", s.elapsed()* 1e3, " ms");
+    if correctness then
+      writeln("Chapel Sum (", function, "): ", + reduce A);
+  }
+
+  cu_nvcc_main(printTime, correctness);
+  cu_clang_main(printTime, correctness);
+
+
+}
+
+inline proc kernel(param function: string, const D, ref A) {
+  @assertOnGpu
+  @gpu.blockSize(256)
+  foreach i in D {
+    A[i] = callFunc(function, i);
+  }
+}
+inline proc callFunc(param function: string, v) {
+  select function {
+    when "tanhf" do return tanh(v:real(32));
```

Comment:
Can we make this repeatedly do `tanh` so as to hide the kernel launch latency by increasing the mathematical intensity?

Reply:
I did that initially as a little loop around `tanh` that sums before setting it in the array. However, this completely changed the performance characteristics and codegen

No Loop:
Chapel: 2.4 ms
nvcc: 1.8 ms
clang: 1.8 ms

With Loop (100 iterations)
Chapel: 18 ms
nvcc: 40 ms
clang: 22 ms

Reply:
OK, I can't stop thinking about it now :)

What happens if you add a loop with 1 trip by default but that can be controlled with a `config const`?

Reply:
I do not know how to explain this.

Before, I had a `config const iterations` in the chapel code and `#define iterations` in the cuda code. This time, I did a single `config const iterations` that gets threaded through to the cuda calls.

With iterations=1, the Chapel code improves to match the cuda code, and the cuda code has no change. With iterations of 100, I get the following:

Chapel: 72ms
nvcc: 24ms
clang: 7ms

Reply:
This is very very weird, but for the sake of commiting this code I have just added the param `useInnerLoop`, along with a note about the weird loop behavior

Reply:
> This is very very weird

Stop trying to entice me!